### DISCUSSION TITLE
Fail-fast for validations

### DISCUSSION THREAD

**Discussion by nirname** [AaronLasseigne/active_interaction#565] [Ideas]
I have got an interaction that does a lot of different things with an email

```ruby
class VerifyEmail < BaseInteraction
  string :email

  validates :email, presence: true, email: true

  validate do
    errors.add(:email, 'please use another domain') unless domain.allowed?
    errors.add(:email, 'user not found') unless user
    errors.add(:email, 'user is blocked') if user.blocked?
  end

  def execute do
    errors.add(:email, 'too many login attempts) if rate_limit_exceeded?
  end
end
```

All the errors scattered across the logic. I'd like to get only the first error, probably, stop all other validations at some point. Fair enough, I can bypass it by asking `errors.messages[:email]` to be `blank?` or calling `return` or `next` at some point.

```ruby
class VerifyEmail < BaseInteraction
  string :email

  validates :email, presence: true, email: true, if: -> { errors.message[:email].blank? }

  validate do
    errors.add(:email, 'please use another domain') && next unless domain.allowed?
    errors.add(:email, 'user not found') && next unless user
    errors.add(:email, 'user is blocked') && next if user.blocked?
  end

  def execute do
    errors.add(:email, 'too many login attempts) and return if rate_limit_exceeded?
  end
end
```

Perhaps this is not even the extension that should be added here, and it is related more to validations mechanism rather than to this gem. But since this idea came to me while I was working with it that is what i'd like to see to reduce the boilerplate:

```ruby
class VerifyEmail < BaseInteraction
  fail_fast :email

  # keep it clean as in the first example
end
```

Even if it is not the right place, may be someone will be inspired by this topic, and implement it somewhere else, or redirect me to the correct place for such a discussion. May be, if it is more or less appropriate I will implement it by myself.

**Comment by antulik** [comment]
I see there are couple of options. 

## 1. use `next` or `return`

Like you suggested

## 2 use `:strict` error mode

```
errors.add(:name, :invalid, strict: true)
```

that will raise `ActiveModel::StrictValidationFailed` which you'll need to rescue in around block, ActiveInteraction does not rescue it.

## 3. use `halt!`

Similar to #1, but this one allows to stop from any sub method.

see for more https://github.com/antulik/active_interaction-extras#halt

**Comment by AaronLasseigne** [comment]
I try to keep it in line with how Rails works. In fact, the validation is straight out of Rails. @antulik gave a good response. I think generally this would be outside of the scope of this gem and more of a Rails question.

Thanks for the question!
